javascript/organizing_your_javascript_code/organizing_code_with_objects.md

```diff
@@ -0,0 +1,374 @@
+
+1. Try modelling the behavior of a piggy bank as an object:
+   1. You should be able to interact with the object via a method that allows you to deposit coins.
+      - For now, use a string to represent the name of a coin. Ex. `piggyBank.deposit('penny')`
+      - You should keep the coin around by storing it in some kind of list.
+   1. The object should also have a property or method that tells you how much money you have saved as a number.
+      - For simplicity, using only the smallest subunit of a given currency is acceptable.
+      - Ex: If there are 5 US quarters in the piggy bank, then the number `125` for '125 cents' is acceptable.
+   1. You should be able to read the list of all coins that are currently in the piggy bank.
+   1. Add a method to the piggy bank object to remove a specific type of coin from the piggy bank if it's available.
+      - Ensure you can still get the correct savings after removing coins!
+   1. Lastly, create a way to remove all the money from the jar and start fresh.
+
+1. Try to model something else as an object! Try to keep it simple. When you're done, consider showing off what you've created on [The Odin Project's Discord Channel](https://discord.com/invite/fbFCkYabZB), and tell us how you're feeling about the idea of organizing code into objects.
+
+#### Extra credit
+
+1. You may have exposed the list that the piggy bank uses to track coins as a public property. Depending on how you implemented the piggy bank, modifying this list without using something like a `deposit` or `withdraw` method on the object could throw some of it's properties out of wack, like the one that tells you the amount of savings you've accrued.
+
+   Additionally, adding an unexpected value to this list by modifying it directly could cause errors within your piggy bank methods. Let's try to ensure that doesn't happen!
+   - Indicate that the list of coins the piggy bank uses is a **private property** of the object.
+   - Create a **public method** that gives the user a *copy* of the list that they can manipulate to their hearts content without breaking the piggy bank object.
+
+1. Try creating a couple 'coin' objects to deposit into the piggy bank instead of using a string.
+   - A coin object might contain properties like the name of the coin, and it's value (preferably in the same unit the piggy bank measures it's savings).
+   - Make sure that your piggy bank object is still able to keep track of savings and remove coins correctly after this change!
+   - After making this change, consider how you are calculating savings for the piggy bank compared to before. Do you prefer working with the objects or the strings more?
+   - Consider how the piggy bank might check for a correct string vs checking for an object with the correct properties when depositing. Which seems easier? Does one seem more flexible than the other?
```

Comment:
I'm not particularly convinced by the value of this extra credit section, especially one that contains so much. Personally, I think the main assignment is absolutely sufficient here. If you feel that anything in the extra credit cannot go, then it ought to be in the main assignment instead of extra credit anyway.

Reply:
Yeah, I felt the main assignment covered "Creating an object with contained functionality well", but it didn't demonstrate things like:
1. The usefulness of grouping data in an object, and passing that around to other functions/methods
2. Using private method or properties as a way to prevent "bad state" in objects. 

I felt like the piggy bank object encapsulated the main point of this lesson, but these extra-credit things could help solidify the importance of the other things touched upon in this lesson. 

I agree the assignment section is unusually long, for an assignment section. The piggy-bank related stuff *could* potentially be a small project following this lesson.

Reply:
I'm afraid I'm running into a problem here where it feels *really* hard let go of a portion of the assignment that gets learners hands-on with private properties.

While the lesson on Factory Functions gives a general overview of how to implement "private variables" with factory functions through closures, I feel it's much more focused on conveying the idea of closures and providing a way to create some form of inheritance with factory functions.

In this lesson I try to create more of a holistic introduction to the concept, so it can be carried over easily to any of the other object creation methods that are covered.

The main article covers using private properties/methods as a way to organize code within an object that doesn't need to be public, and last part of the assignment section tries to tackle instances where private properties can help prevent "bad state". 

I did take a stab at rewriting the assignment section to remove the "Extra Credit" section. I landed at a version that splits creating a model of the piggy bank into three parts, with each part corresponding to a main idea in this lesson.

The first part includes creating a couple very simple coin objects, as practice for grouping related data together:
```
1. Try modelling the behavior of a piggy bank as an object
   1. Create a few 'coin' objects, and assign them to variables to deposit in the piggy bank later.
     - A coin object should contain a `name` property, containing the name of the coin, and a `value` property that give's it's value as a number.
     - For simplicity, using only the smallest subunit of a given currency is acceptable to convey the coin's value.

          Ex: One US quarter may have a value of `25` to represent '25 cents'.
```

The second focuses on the actual piggy bank, and organizing data and functionality together:
```
   1. Now, create the piggy bank object:
      1. There should be a public `savings` or `getSavings` property or method that tells users of the piggy bank how much money it contains, as a number. This should give the total value of all the coins currently stored in the piggy bank.
      1. You should be able to interact with your piggy bank object through a public `deposit` method, allowing you to deposit coins. This function should accept one of your coin objects as an argument, and keep it around by storing it in some kind of list.
        - Ensure the piggy bank's savings is updated when you add a coin!
      1. You should be able to get a list of all the coins that are currently stored in the piggy bank through a `coins`/`getCoins` property or method.
      1. Add a `withdraw` method that allows you to remove a specific type of coin from the piggy bank if it's available.
        - Ensure the piggy bank's savings shows the correct value after removing coins!
        - This method should return the coin object that it removes from the list.
      1. Create a method to remove all the money from the jar and start fresh.
      1. You should now have a functional piggy bank object that you can play with in the browser console!
```

And the third provides practice for potentially common scenarios where you might want to use private methods/properties to avoid bad state:
```
   1. Lastly, make sure the data on your piggy bank object can't be interacted with in ways you didn't intend! This is an area where private properties/methods come in handy.
      1. You should have exposed the list that the piggy bank uses to track coins through a public property or method. Depending on how you implemented this functionality, modifying this list *from the outside*, without using something like a `deposit` or `withdraw` method could result in coins being added or removed without the savings being updated!

         Additionally, adding an unexpected value to this list by modifying it directly could cause errors within your piggy bank methods (what if somebody adds something that's not a coin object?). Let's try to ensure that doesn't happen!

          - Indicate that the list of coins the piggy bank uses is a **private property** of the object.
          - Create a **public method** that gives the user a *copy* of the list and it's objects that they can manipulate to their hearts content without breaking the piggy bank object.          - If a user of your object tries to deposit something that doesn't appear to be a coin with the `deposit` method, throw an error!

      1. Did you use a public property to convey the savings to the consumer of your object? That property can probably be modified willy-nilly to show the wrong value! There are a couple ways you could fix this:
         - Use a public method that calculates the savings every time it's called
         - Another option is to create a private property tracking the savings, and use a public method to return it's value.
```

I think ultimately this becomes a part of the bigger conversation we're having about how much focus private properties/methods should have at this point. I would personally lean towards approaching the topic earlier, then explaining how one might implement them with each object creation method. 

If it's decided it's best to not explore them earlier than we already are, I still feel like there's probably improvements to how we approach the concept later. Maybe it's just my experience and bias, but this feels like a concept that's people grasp understanding of more through discussion in the server than through the lessons themselves.